File: leetcode/04-Zigzag-Conversion/python/main.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        # numRowsが１の時はそのまま返す
        if numRows == 1:
            return s

        # ジグザグの表を作る
        table = [[] for _ in range(numRows)]
        currentRow = 0
        goDown = True

        for i in range(len(s)):
            table[currentRow].append(s[i])

            if goDown:
                currentRow += 1
            else:
                currentRow -= 1

            if currentRow == 0:
                goDown = True
            elif currentRow == numRows - 1:
                goDown = False

        # 繋げて文字列にする
        return ''.join([''.join(x) for x in table])

File: leetcode/04-Zigzag-Conversion/go/main.go
package main

import (
	"strings"
)

func convert(s string, numRows int) string {
	if len(s) == 1 || numRows == 1 {
		return s
	}

	table := make([][]string, numRows)
	currentRow := 0
	goDown := true

	for i := 0; i < len(s); i++ {
		currentChar := s[i : i+1]
		table[currentRow] = append(table[currentRow], currentChar)

		if currentRow == 0 {
			goDown = true
		} else if currentRow == numRows-1 {
			goDown = false
		}

		if goDown {
			currentRow += 1
		} else {
			currentRow -= 1
		}
	}

	tmpList := []string{}
	for i := 0; i < len(table); i++ {
		tmpList = append(tmpList, strings.Join(table[i], ""))
	}

	return strings.Join(tmpList, "")

}
